This looks like a trivial task, but I have very limited Rust knowledge.
So when I use `Url` instead of a `String`, I am getting these errors:

```
103 | #[derive(Encode, Decode, Serialize, Deserialize, Debug, Clone, Eq, PartialEq)]
    |          ^^^^^^ the trait `minicbor::Encode<Ctx>` is not implemented for `Url`
    |
    = help: the following other types implement trait `minicbor::Encode<C>`:
              <&T as minicbor::Encode<C>>
              <&mut T as minicbor::Encode<C>>
              <() as minicbor::Encode<C>>
              <(A, B) as minicbor::Encode<Ctx>>
              <(A, B, C) as minicbor::Encode<Ctx>>
              <(A, B, C, D) as minicbor::Encode<Ctx>>
              <(A, B, C, D, E) as minicbor::Encode<Ctx>>
              <(A, B, C, D, E, F) as minicbor::Encode<Ctx>>
            and 233 others

error[E0277]: the trait bound `Url: minicbor::Decode<'_, Ctx>` is not satisfied
   --> implementations/rust/ockam/ockam_api/src/cloud/project.rs:103:18
    |
103 | #[derive(Encode, Decode, Serialize, Deserialize, Debug, Clone, Eq, PartialEq)]
    |                  ^^^^^^ the trait `minicbor::Decode<'_, Ctx>` is not implemented for `Url`
    |
    = help: the following other types implement trait `minicbor::Decode<'b, C>`:
              <&'a ByteSlice as minicbor::Decode<'b, C>>
              <&'a std::path::Path as minicbor::Decode<'b, C>>
              <&'a str as minicbor::Decode<'b, C>>
              <() as minicbor::Decode<'b, C>>
              <(A, B) as minicbor::Decode<'b, Ctx>>
              <(A, B, C) as minicbor::Decode<'b, Ctx>>
              <(A, B, C, D) as minicbor::Decode<'b, Ctx>>
              <(A, B, C, D, E) as minicbor::Decode<'b, Ctx>>
            and 227 others

```

So I know that `Encoder` and `Decoder` are missing for `Url`, but not sure what is the idiomatic way to deal with this.
There is no way to call the `derive` macro on an external type like `Url`, and I was not able to find a similar example in the codebase. @etorreborre

--- a/implementations/rust/ockam/ockam_api/Cargo.toml
+++ b/implementations/rust/ockam/ockam_api/Cargo.toml
@@ -55,6 +55,7 @@ time = { version = "0.3.23", default-features = false }
 tinyvec = { version = "1.6.0", features = ["rustc_1_57"] }
 tokio-retry = "0.3.0"
 tracing = { version = "0.1", default-features = false }
+url = "2.4.0"
 
 ockam = { path = "../ockam", version = "^0.90.0", features = ["software_vault"] }
 ockam_multiaddr = { path = "../ockam_multiaddr", version = "0.24.0", features = ["cbor", "serde"] }

--- a/implementations/rust/ockam/ockam_api/src/cloud/project.rs
+++ b/implementations/rust/ockam/ockam_api/src/cloud/project.rs
@@ -13,6 +13,7 @@ use ockam_multiaddr::MultiAddr;
 use ockam_node::tokio;
 
 use crate::error::ApiError;
+use crate::minicbor_url::Url;
 
 #[derive(Encode, Decode, Serialize, Deserialize, Debug, Default, Clone, Eq, PartialEq)]
 #[cbor(map)]
@@ -107,7 +108,7 @@ pub struct OktaConfig {
     #[serde(skip)]
     #[cbor(b(0))] pub tag: TypeTag<6434814>,
 
-    #[cbor(b(1))] pub tenant_base_url: String,
+    #[cbor(b(1))] pub tenant_base_url: Url,
 
     #[cbor(b(2))] pub certificate: String,
 
@@ -118,7 +119,7 @@ pub struct OktaConfig {
 
 impl<'a> OktaConfig {
     pub fn new<S: ToString, T: AsRef<str>>(
-        tenant_base_url: S,
+        tenant_base_url: Url,
         certificate: S,
         client_id: S,
         attributes: &'a [T],
@@ -126,7 +127,7 @@ impl<'a> OktaConfig {
         Self {
             #[cfg(feature = "tag")]
             tag: TypeTag,
-            tenant_base_url: tenant_base_url.to_string(),
+            tenant_base_url,
             certificate: certificate.to_string(),
             client_id: client_id.to_string(),
             attributes: attributes.iter().map(|x| x.as_ref().to_string()).collect(),
@@ -134,14 +135,14 @@ impl<'a> OktaConfig {
     }
 
     pub fn new_empty_attributes<S: ToString>(
-        tenant_base_url: S,
+        tenant_base_url: Url,
         certificate: S,
         client_id: S,
     ) -> Self {
         Self {
             #[cfg(feature = "tag")]
             tag: TypeTag,
-            tenant_base_url: tenant_base_url.to_string(),
+            tenant_base_url,
             certificate: certificate.to_string(),
             client_id: client_id.to_string(),
             attributes: Vec::new(),
@@ -151,7 +152,7 @@ impl<'a> OktaConfig {
 
 #[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
 pub struct OktaAuth0 {
-    pub tenant_base_url: String,
+    pub tenant_base_url: Url,
     pub client_id: String,
     pub certificate: String,
 }

--- a/implementations/rust/ockam/ockam_api/src/config/lookup.rs
+++ b/implementations/rust/ockam/ockam_api/src/config/lookup.rs
@@ -243,7 +243,7 @@ impl ProjectLookup {
         )
         .await?;
         let okta = project.okta_config.as_ref().map(|o| OktaAuth0 {
-            tenant_base_url: o.tenant_base_url.to_string(),
+            tenant_base_url: o.tenant_base_url.clone(),
             client_id: o.client_id.to_string(),
             certificate: o.certificate.to_string(),
         });

--- a/implementations/rust/ockam/ockam_api/src/lib.rs
+++ b/implementations/rust/ockam/ockam_api/src/lib.rs
@@ -134,6 +134,7 @@ pub mod error;
 pub mod hop;
 pub mod identity;
 pub mod kafka;
+pub mod minicbor_url;
 pub mod nodes;
 pub mod okta;
 pub mod port_range;

--- a/implementations/rust/ockam/ockam_api/src/minicbor_url.rs
+++ b/implementations/rust/ockam/ockam_api/src/minicbor_url.rs
@@ -1,0 +1,79 @@
+use minicbor::{decode, encode, Decode, Decoder, Encode, Encoder};
+use serde::{de, Deserialize, Deserializer, Serialize, Serializer};
+use std::fmt::{Display, Formatter};
+use std::ops::Deref;
+use std::str::FromStr;
+use url::{ParseError, Url as RegularUrl};
+
+/// Adding newtype for url::Url, since we can't use derive macro for external types
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct Url(RegularUrl);
+
+impl Url {
+    pub fn new(url: RegularUrl) -> Self {
+        Self(url)
+    }
+
+    pub fn parse(input: &str) -> Result<Url, ParseError> {
+        let url = RegularUrl::parse(input)?;
+        Ok(Url(url))
+    }
+}
+
+impl Deref for Url {
+    type Target = RegularUrl;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl Display for Url {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.write_str(self.0.as_str())
+    }
+}
+
+impl<C> Encode<C> for Url {
+    fn encode<W: encode::Write>(
+        &self,
+        e: &mut Encoder<W>,
+        ctx: &mut C,
+    ) -> Result<(), encode::Error<W::Error>> {
+        self.0.as_str().encode(e, ctx)
+    }
+}
+
+impl<'b, C> Decode<'b, C> for Url {
+    fn decode(d: &mut Decoder<'b>, ctx: &mut C) -> Result<Self, decode::Error> {
+        let url = d.decode_with(ctx)?;
+        let url = RegularUrl::from_str(url).map_err(decode::Error::message)?;
+        Ok(Url(url))
+    }
+}
+
+impl Serialize for Url {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        serializer.serialize_str(self.0.as_str())
+    }
+}
+
+impl<'d> Deserialize<'d> for Url {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'d>,
+    {
+        let url: &str = Deserialize::deserialize(deserializer)?;
+        let url = RegularUrl::from_str(url).map_err(de::Error::custom)?;
+        Ok(Url(url))
+    }
+}
+
+impl From<Url> for RegularUrl {
+    fn from(url: Url) -> Self {
+        url.0
+    }
+}

--- a/implementations/rust/ockam/ockam_command/src/project/addon/configure_okta.rs
+++ b/implementations/rust/ockam/ockam_command/src/project/addon/configure_okta.rs
@@ -6,7 +6,6 @@ use clap::builder::NonEmptyStringValueParser;
 use clap::Args;
 use colorful::Colorful;
 use miette::{miette, Context as _, IntoDiagnostic};
-use reqwest::Url;
 use rustls::{Certificate, ClientConfig, ClientConnection, Connection, RootCertStore, Stream};
 
 use ockam::Context;
@@ -14,6 +13,7 @@ use ockam_api::cli_state::{StateDirTrait, StateItemTrait};
 use ockam_api::cloud::operation::CreateOperationResponse;
 use ockam_api::cloud::project::{OktaConfig, Project};
 use ockam_api::cloud::CloudRequestWrapper;
+use ockam_api::minicbor_url::Url;
 use ockam_core::api::Request;
 use ockam_core::CowStr;
 
@@ -119,7 +119,7 @@ async fn run_impl(
         _ => query_certificate_chain(domain)?,
     };
 
-    let okta_config = OktaConfig::new(tenant, certificate, client_id, &attributes);
+    let okta_config = OktaConfig::new(base_url, certificate, client_id, &attributes);
     let body = okta_config.clone();
 
     // Validate okta configuration

